sys/net/application_layer/gcoap/fileserver.c

```diff
@@ -205,25 +226,194 @@ static ssize_t gcoap_fileserver_file_handler(coap_pkt_t *pdu, uint8_t *buf, size
     return coap_get_total_hdr_len(pdu);
 }
 
-static ssize_t gcoap_fileserver_directory_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len,
-                                                  struct requestdata *request,
-                                                  gcoap_fileserver_entry_t *resource)
+static ssize_t _put_file(coap_pkt_t *pdu, uint8_t *buf, size_t len,
+                         struct requestdata *request,
+                         gcoap_fileserver_entry_t *resource)
+{
+    (void)resource;
+    int ret;
+    uint32_t etag;
+    struct stat stat;
+    coap_block1_t block1 = {0};
+    bool create = (vfs_stat(request->namebuf, &stat) == -ENOENT);
+    if (create) {
+        /* While a file 'f' is initially being created,
+           we save the partial content in '.f' and rename it afterwards */
+        if (!(ret = strlen(request->namebuf)) || (unsigned)ret >= sizeof(request->namebuf) - 1) {
+            /* need one more char '.' */
+            return gcoap_fileserver_error_handler(pdu, buf, len, -ENOBUFS);
+        }
+        while (--ret) {
+            if (request->namebuf[ret] == '/') {
+                memmove(&request->namebuf[ret + 2],
+                        &request->namebuf[ret + 1],
+                        strlen(&request->namebuf[ret + 1]));
+                request->namebuf[ret + 1] = '.';
+                break;
+            }
+        }
```

Comment:
This needs some explaining.
Why not use `strrchr(request->namebuf, '/')`?

Reply:
I was not aware of `strrchr()`. But it does what you think it does: move the file component one char to the right and put a '.' in front of it.